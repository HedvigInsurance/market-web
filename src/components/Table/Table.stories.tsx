import React from 'react'
import { colorsV3 } from '@hedviginsurance/brand'
import { Tick } from '../icons/Tick'
import {
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableProps,
  TableRow,
} from './Table'

export default {
  title: 'Components/Table',
  component: Table,
  parameters: {
    paddings: [
      { name: 'Medium', value: '32px' },
      { name: 'Large', value: '64px', default: true },
    ],
  },
}

function createDataCarInsuranceData(
  name: string,
  traffic: boolean,
  half: boolean,
  full: boolean,
) {
  return { name, traffic, half, full }
}

const comparisonRows = [
  createDataCarInsuranceData('Personskador', true, true, true),
  createDataCarInsuranceData('Skador på annans egendom', true, true, true),
  createDataCarInsuranceData('Stöld och inbrott', false, true, true),
  createDataCarInsuranceData('Brand', false, true, true),
  createDataCarInsuranceData('Glasskador', false, true, true),
  createDataCarInsuranceData('Bärgning', false, true, true),
  createDataCarInsuranceData('Skador på bilen från olycka', false, false, true),
]

const premiumRows = [
  createDataCarInsuranceData('Personskador', false, true, true),
  createDataCarInsuranceData('Skador på annans egendom', false, true, true),
]

const tableProps: TableProps = {
  minWidth: '650px',
}

export const Default = () => (
  <>
    <Table {...tableProps}>
      <TableHead>
        <TableRow>
          <TableCell></TableCell>
          <TableCell align="center">Trafik</TableCell>
          <TableCell align="center">Halv</TableCell>
          <TableCell align="center">Hel</TableCell>
        </TableRow>
      </TableHead>
      <TableBody>
        {comparisonRows.map((row) => (
          <TableRow key={row.name}>
            <TableCell>{row.name}</TableCell>
            <TableCell
              align="center"
              color={row.traffic ? undefined : colorsV3.gray400}
            >
              {row.traffic ? <Tick /> : <>&mdash;</>}
            </TableCell>
            <TableCell
              align="center"
              color={row.half ? undefined : colorsV3.gray400}
            >
              {row.half ? <Tick /> : <>&mdash;</>}
            </TableCell>
            <TableCell
              align="center"
              color={row.full ? undefined : colorsV3.gray400}
            >
              {row.full ? <Tick /> : <>&mdash;</>}
            </TableCell>
          </TableRow>
        ))}
      </TableBody>
    </Table>
    <Table {...tableProps} mt="2rem">
      <TableHead>
        <TableRow>
          <TableCell>ENDAST MED PREMIUM</TableCell>
        </TableRow>
      </TableHead>
      <TableBody>
        {premiumRows.map((row) => (
          <TableRow key={row.name}>
            <TableCell>{row.name}</TableCell>
            <TableCell
              align="center"
              color={row.traffic ? undefined : colorsV3.gray400}
            >
              {row.traffic ? <Tick /> : <>&mdash;</>}
            </TableCell>
            <TableCell
              align="center"
              color={row.half ? undefined : colorsV3.gray400}
            >
              {row.half ? <Tick /> : <>&mdash;</>}
            </TableCell>
            <TableCell
              align="center"
              color={row.full ? undefined : colorsV3.gray400}
            >
              {row.full ? <Tick /> : <>&mdash;</>}
            </TableCell>
          </TableRow>
        ))}
      </TableBody>
    </Table>
  </>
)
