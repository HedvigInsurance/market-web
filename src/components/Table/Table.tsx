import styled from '@emotion/styled'
import React, { createContext, ReactNode, useContext } from 'react'
import { colorsV3 } from '@hedviginsurance/brand'
import { getMargins, Margins } from '../../utils/margins'
import { TABLET_BP_UP } from '../blockHelpers'

type WithChildrenProps = {
  children?: ReactNode
}

export type TableProps = {
  minWidth?: string
  maxWidth?: string
  fullWidth?: boolean
} & Margins

const Table = styled.table(
  ({ minWidth, maxWidth, fullWidth, ...props }: TableProps) => ({
    borderCollapse: 'collapse',
    borderSpacing: 0,
    minWidth,
    maxWidth,
    width: fullWidth ? '100%' : 'auto',
    ...getMargins(props),
  }),
)

const TableHeadElement = styled.thead()
// This context is used to conditionally render a `th` inside `thead` elements, and `td` otherwise
const TableHeadContext = createContext(false)

const TableHead = ({ children }: WithChildrenProps) => {
  return (
    <TableHeadContext.Provider value={true}>
      <TableHeadElement>{children}</TableHeadElement>
    </TableHeadContext.Provider>
  )
}

const TableRow = styled.tr()

const TableBody = styled.tbody()

const getTableCellStyles = ({ align = 'left', color }: TableCellProps) => ({
  textAlign: align,
  color,
})

type TableCellProps = WithChildrenProps & {
  align?: 'center' | 'left' | 'right'
  color?: keyof typeof colorsV3 | undefined | string
}

const TableCellElement = styled.td((props: TableCellProps) => ({
  ...getTableCellStyles(props),
  padding: '0.5rem 0.5rem',

  [TABLET_BP_UP]: {
    padding: '0.5rem 1rem',
  },

  borderBottom: `1px solid ${colorsV3.gray300}`,
}))

const TableHeadCellElement = styled.th(
  (props: Omit<TableCellProps, 'color'>) => ({
    ...getTableCellStyles(props),
    padding: '1rem 0.5rem',
    fontSize: '0.875rem',
    fontWeight: 'normal',
    color: colorsV3.gray700,

    [TABLET_BP_UP]: {
      padding: '1rem',
    },
  }),
)

const TableCell = ({ children, ...props }: TableCellProps) => {
  const context = useContext(TableHeadContext)

  // Render a `th` when inside a `thead` and `td` otherwise
  if (context) {
    return <TableHeadCellElement {...props}>{children}</TableHeadCellElement>
  }

  return <TableCellElement {...props}>{children}</TableCellElement>
}

export { Table, TableHead, TableRow, TableCell, TableBody }
