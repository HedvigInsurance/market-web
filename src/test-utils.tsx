import React from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { Provider } from 'constate'
import { HelmetProvider } from 'react-helmet-async'
import { v4 as uuidV4 } from 'uuid'
import { BrowserRouter } from 'react-router-dom'
import { locales } from 'utils/locales'
import { LocaleProvider } from './context/LocaleContext'

const initialState = {
  story: {
    story: {
      uuid: uuidV4(),
      name: 'blargh',
      full_slug: '/home',
      content: { _uid: uuidV4(), body: [] },
    },
  },
}

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <Provider initialState={initialState}>
      <LocaleProvider currentLocale={locales['se-en']}>
        <BrowserRouter>
          <HelmetProvider>{children}</HelmetProvider>
        </BrowserRouter>
      </LocaleProvider>
    </Provider>
  )
}

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: AllTheProviders, ...options })

export * from '@testing-library/react'
export { customRender as renderComponent }
