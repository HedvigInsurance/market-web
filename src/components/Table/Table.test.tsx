import React from 'react'
import { renderComponent } from '../../test-utils'
import { Table, TableBody, TableCell, TableHead, TableRow } from './Table'

describe('Table', () => {
  it('can render', () => {
    const { container } = renderComponent(
      <Table>
        <TableBody>
          <TableRow>
            <TableCell>hello</TableCell>
          </TableRow>
        </TableBody>
      </Table>,
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('can render table header', () => {
    const { container } = renderComponent(
      <Table>
        <TableHead>
          <TableRow>
            <TableCell>hello</TableCell>
          </TableRow>
        </TableHead>
      </Table>,
    )

    expect(container.firstChild).toMatchSnapshot()
  })
})
