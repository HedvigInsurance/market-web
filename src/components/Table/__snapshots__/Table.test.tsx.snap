// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table can render 1`] = `
<table
  class="css-qfk35h-Table e1vwxjz80"
>
  <tbody
    class="css-r22c21-TableBody e1vwxjz83"
  >
    <tr
      class="css-168pto2-TableRow e1vwxjz82"
    >
      <td
        class="css-1etym39-TableCellElement e1vwxjz84"
      >
        hello
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table can render table header 1`] = `
<table
  class="css-qfk35h-Table e1vwxjz80"
>
  <thead
    class="css-1q4jspl-TableHeadElement e1vwxjz81"
  >
    <tr
      class="css-168pto2-TableRow e1vwxjz82"
    >
      <th
        class="css-1pp02jo-TableHeadCellElement e1vwxjz85"
      >
        hello
      </th>
    </tr>
  </thead>
</table>
`;
